Harden error handling across Thunderstore UI

Oksamies opened the pull request:
- document the end-to-end error-handling initiative, plans, and audit outcomes under docs
- add user-facing error primitives to @thunderstore/thunderstore-api, sanitize server detail, and cover the new behaviour with unit tests
- streamline @thunderstore/dapper-ts so its methods return raw Thunderstore API payloads without reshaping
- expose consistent UserFacingError handling through useApiCall, ApiAction, useFormToaster, and useStrongForm
- introduce the Nimbus error boundary toolkit plus shared loader helpers, then wire the root layout into the new fallbacks
- wrap community, package, settings, and upload routes with Nimbus boundaries, improve loader error mapping, and enhance Suspense fallbacks

Oksamies committed:
feat(thunderstore-api): normalize user-facing errors

Introduce the sanitized ApiError helpers, payload parsing, and UserFacingError mapping, update apiFetch to use them, streamline the get/post/delete wrappers, and add focused tests for the new messaging.

Oksamies committed:
refactor(dapper): return raw API payloads

Update PaginatedList typing and the dapper-ts methods to stop re-shaping results and instead forward the Thunderstore API responses directly.

Oksamies committed:
feat(forms): surface mapped API errors

Teach useApiCall, ApiAction, useFormToaster, and useStrongForm to emit UserFacingError instances so UI toasts and handlers receive consistent copy.

Oksamies committed:
feat(ui): add Nimbus error boundary toolkit

Land the shared error utilities (handleLoaderError, payload resolvers, NimbusErrorBoundary, getLoaderTools) and wire the root layout into the new boundary and fallback system.

Oksamies committed:
feat(ui): harden community browsing

Adopt the new loader tooling across the community routes and package search experience, including revised CSS and Suspense fallbacks.

Oksamies committed:
feat(ui): harden package surfaces

Apply mapped loader errors and boundary coverage to package listing, tabs, dependant/teams views, and shared components such as paginated dependencies and report form.

Oksamies committed:
refactor(team): use loader tools, stream filters+listings, and add suspense/error boundaries

- replace manual DapperTs/session setup with getLoaderTools()
- return filtersAndListings (Promise of [filters, listings]) from loaders to enable streaming
- wrap server loader in try/catch and map errors via handleLoaderError/createNotFoundMapping
- use throwUserFacingPayloadResponse for missing params instead of raw Response
- update Team UI to Suspense + Await with SkeletonBox and Nimbus error element; add Nimbus route ErrorBoundary

Oksamies committed:
feat(ui): harden settings & upload flows

Wrap the user/team settings and upload routes with Nimbus boundaries, rely on the new loader helpers, and ensure form toasts use the user-facing error formatter.

Oksamies committed:
feat(ui): parse & display user-facing error payloads in root ErrorBoundary; add NimbusAwaitErrorElement to community Await

Oksamies committed:
FIx: enhance community component to utilize VITE_BETA_SITE_URL for dynamic OG URL generation

Oksamies committed:
feat(utils): add parseIntegerSearchParam function for safe integer parsing; update usages across components

Oksamies committed:
fix(utils): correct variable usage in getLoaderTools function for session tools

Oksamies committed:
fix(errors): improve error handling by adding optional chaining for options in handleLoaderError function

Oksamies committed:
fix(types): correct default type for SubmissionError in UseStrongFormProps interface

Oksamies committed:
feat(dapper-fake): enhance getFakePackageListings to support pagination and dynamic query parameters

Oksamies committed:
fix(types): enforce SubmissionDataShape type constraint in useStrongForm interfaces and add input validation

Oksamies committed:
feat(errors): refactor NimbusErrorBoundaryFallback to use classnames utility for dynamic class assignment

Oksamies committed:
style(community): standardize spacing and padding using CSS variables

Oksamies committed:
fix(package): improve string interpolation for package dependants API endpoint

Oksamies committed:
docs(errors): add error handling guide covering handleLoaderError, UserFacingError, and NimbusErrorBoundary with usage examples

anttimaki commented on docs/error-handling.md:
```diff
@@ -0,0 +1,242 @@
+
+This can lead to less specific error messages for the user. The recommendation is to apply the same `try...catch` and `handleLoaderError` pattern in `clientLoader` as is used in `loader` to ensure a consistent user experience.
+
+#### Bad Example (`clientLoader` without `handleLoaderError`)
```
I would probably drop the bad example and focus on good example(s).

anttimaki commented on docs/error-handling.md:
```diff
@@ -0,0 +1,242 @@
+    // Awaiting all follow-up data forces the UI to block on the clientLoader.
+    const members = await dapper.getTeamMembers(params.teamId);
+    return { team, members };
+  } catch (error) {
```
(Don't catch an error just to throw it again.)

anttimaki commented on docs/error-handling.md:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  // Kick off dependent work without awaiting so Suspense can resolve it downstream.
+  const members = dapper.getTeamMembers(team.slug);
```
Should this be inside the try-block above? If it's intentionally left out, maybe use the comment above to describe what happens if it's erroneously placed inside the try-block.

anttimaki commented on docs/error-handling.md:
```diff
@@ -0,0 +1,242 @@
+  return (
+    <NimbusErrorBoundaryFallback
+      error={error}
+      title={payload.headline}
```
Could payload.headline be renamed to payload.title so it's in line with usage (and also shorter).

anttimaki commented on docs/error-handling.md:
````diff
@@ -0,0 +1,242 @@
+When handing promises to the UI, wrap the consuming view with `Suspense`/`Await` and provide `NimbusAwaitErrorElement` as the `errorElement`. This ensures promise rejections render the same Nimbus fallback instead of crashing the tree.
+
+```tsx
+// apps/cyberstorm-remix/app/p/tabs/Readme/Readme.tsx
````
This is not what's done in the readme component (copy-paste comment?).

anttimaki reviewed (commented):
This review is only for the `error-handing.md` file. The main points are that I would attempt to improve the clarity by restructuring the document:

1. Reading the document with my level of knowledge of the project and the technologies used, I think structuring the doc based on the "core concepts" isn't the clearest way to do it. I would instead structure it to 1. Server side loader 2. Client side clientLoader 3. Error boundaries

2. For error boundaries, it's not perfectly clear to me what's the point of separate sections "## React Router Related Error Handling" and "## `NimbusErrorBoundary`"? Could we have just one section that tells us how to do things in Cyberstorm Remix app? I.e. combine the two parts to a structure something like 1. Root level error boundary ("be aware it exist, you don't need to worry about it") 2. Route-level Error boundary (including using suspense?) in your component 3. Granular Component Boundary in your component

Oksamies committed:
Refactor loaders, error handling, and API helpers per PR review

Dapper / API Helpers
- Removed redundant async keywords from dapper-ts methods (team, communities, ratedPackages).
- Fixed apiFetch generics in thunderstore-api to correctly resolve SchemaOrUndefined.
- Updated dapper-fake pagination to return next/previous URLs and normalize page numbers.

Remix Loaders & Error Handling
- Updated Required tab loaders to use params.packageVersion instead of latest.
- Added parseIntegerSearchParam guards for page parameters across loaders.
- Standardized error handling in Dependants client loader.
- Enabled streaming for Members tab loader.
- Restored server loaders for Profile and TeamSettings to ensure consistent error surfacing.
- Enhanced handleLoaderError to merge default mappings even when options are undefined.
- Updated getLoaderTools to return sessionTools on the client side.

StrongForm & Components
- Improved useStrongForm type safety: enforced refiner when shapes differ, aligned SubmissionError constraints, and safer error casting.
- Fixed stale useMemo dependencies in packageVersion and packageListing to support intra-route navigation.

Documentation
- Restructured error-handling.md to clarify server vs. client loader patterns and error boundary usage.

Oksamies commented:
Closing as too stale
